File: user_app/lib/models/token.dart
class TokenFeedback {
  const TokenFeedback({
    this.rating,
    this.comment,
    this.submittedAt,
  });

  final int? rating;
  final String? comment;
  final DateTime? submittedAt;

  factory TokenFeedback.fromJson(Map<String, dynamic>? json) {
    if (json == null) return const TokenFeedback();
    return TokenFeedback(
      rating: (json['rating'] as num?)?.toInt(),
      comment: json['comment']?.toString(),
      submittedAt: json['submittedAt'] != null ? DateTime.tryParse(json['submittedAt'].toString()) : null,
    );
  }

  Map<String, dynamic> toJson() => {
        'rating': rating,
        'comment': comment,
      };
}

class TokenModel {
  const TokenModel({
    required this.id,
    required this.tokenCode,
    required this.tokenNumber,
    required this.userId,
    required this.centerId,
    required this.serviceId,
    required this.status,
    this.centerName,
    this.serviceName,
    this.counterId,
    this.counterName,
    this.counterNumber,
    this.initialPosition,
    this.currentPosition,
    this.waitEstimateMinutes,
    this.qrData,
    this.calledAt,
    this.servingAt,
    this.completedAt,
    this.createdAt,
    this.actualServiceSeconds,
    this.feedback,
  });

  final String id;
  final String tokenCode;
  final int tokenNumber;
  final String userId;
  final String centerId;
  final String serviceId;
  final String status;
  final String? centerName;
  final String? serviceName;
  final String? counterId;
  final String? counterName;
  final int? counterNumber;
  final int? initialPosition;
  final int? currentPosition;
  final int? waitEstimateMinutes;
  final String? qrData;
  final DateTime? calledAt;
  final DateTime? servingAt;
  final DateTime? completedAt;
  final DateTime? createdAt;
  final int? actualServiceSeconds;
  final TokenFeedback? feedback;

  bool get isActive => ['WAITING', 'CALLED', 'SERVING'].contains(status.toUpperCase());
  bool get isWaiting => status.toUpperCase() == 'WAITING';
  bool get isCalled => status.toUpperCase() == 'CALLED';
  bool get isServing => status.toUpperCase() == 'SERVING';
  bool get isCompleted => status.toUpperCase() == 'COMPLETED';
  bool get canCancel => status.toUpperCase() == 'WAITING';
  bool get hasFeedback => feedback?.rating != null;

  factory TokenModel.fromJson(Map<String, dynamic> json) {
    // Service name extraction
    String sName = 'Service';
    String sId = '';
    final sObj = json['serviceId'];
    if (sObj is Map) {
      sId = (sObj['_id'] ?? sObj['id'] ?? '').toString();
      sName = (sObj['name'] ?? 'Service').toString();
    } else if (sObj != null) {
      sId = sObj.toString();
    }

    // Center name extraction
    String cName = 'Service Center';
    String cId = '';
    final cObj = json['centerId'];
    if (cObj is Map) {
      cId = (cObj['_id'] ?? cObj['id'] ?? '').toString();
      cName = (cObj['name'] ?? 'Service Center').toString();
    } else if (cObj != null) {
      cId = cObj.toString();
    }

    // Counter info extraction
    String? cntId;
    String? cntName;
    int? cntNum;
    final cntObj = json['counterId'];
    if (cntObj is Map) {
      cntId = (cntObj['_id'] ?? cntObj['id'] ?? '').toString();
      cntName = cntObj['name']?.toString();
      cntNum = (cntObj['number'] as num?)?.toInt();
    } else if (cntObj != null) {
      cntId = cntObj.toString();
    }

    return TokenModel(
      id: (json['_id'] ?? json['id'] ?? '').toString(),
      tokenCode: (json['tokenCode'] ?? '').toString(),
      tokenNumber: (json['tokenNumber'] as num?)?.toInt() ?? 0,
      userId: (json['userId'] is Map ? json['userId']['_id'] : json['userId'] ?? '').toString(),
      centerId: cId,
      centerName: cName,
      serviceId: sId,
      serviceName: sName,
      status: json['status']?.toString() ?? 'UNKNOWN',
      counterId: cntId,
      counterName: cntName,
      counterNumber: cntNum,
      initialPosition: (json['initialPosition'] as num?)?.toInt(),
      currentPosition: (json['currentPosition'] as num?)?.toInt(),
      waitEstimateMinutes: (json['waitEstimateMinutes'] as num?)?.toInt(),
      qrData: json['qrData']?.toString(),
      calledAt: json['calledAt'] != null ? DateTime.tryParse(json['calledAt'].toString()) : null,
      servingAt: json['servingAt'] != null ? DateTime.tryParse(json['servingAt'].toString()) : null,
      completedAt: json['completedAt'] != null ? DateTime.tryParse(json['completedAt'].toString()) : null,
      createdAt: json['createdAt'] != null ? DateTime.tryParse(json['createdAt'].toString()) : null,
      actualServiceSeconds: (json['actualServiceSeconds'] as num?)?.toInt(),
      feedback: json['feedback'] is Map ? TokenFeedback.fromJson(Map<String, dynamic>.from(json['feedback'])) : null,
    );
  }

  TokenModel copyWith({
    String? id,
    String? tokenCode,
    int? tokenNumber,
    String? userId,
    String? centerId,
    String? serviceId,
    String? status,
    String? centerName,
    String? serviceName,
    String? counterId,
    String? counterName,
    int? counterNumber,
    int? initialPosition,
    int? currentPosition,
    int? waitEstimateMinutes,
    String? qrData,
    DateTime? calledAt,
    DateTime? servingAt,
    DateTime? completedAt,
    DateTime? createdAt,
    int? actualServiceSeconds,
    TokenFeedback? feedback,
  }) {
    return TokenModel(
      id: id ?? this.id,
      tokenCode: tokenCode ?? this.tokenCode,
      tokenNumber: tokenNumber ?? this.tokenNumber,
      userId: userId ?? this.userId,
      centerId: centerId ?? this.centerId,
      serviceId: serviceId ?? this.serviceId,
      status: status ?? this.status,
      centerName: centerName ?? this.centerName,
      serviceName: serviceName ?? this.serviceName,
      counterId: counterId ?? this.counterId,
      counterName: counterName ?? this.counterName,
      counterNumber: counterNumber ?? this.counterNumber,
      initialPosition: initialPosition ?? this.initialPosition,
      currentPosition: currentPosition ?? this.currentPosition,
      waitEstimateMinutes: waitEstimateMinutes ?? this.waitEstimateMinutes,
      qrData: qrData ?? this.qrData,
      calledAt: calledAt ?? this.calledAt,
      servingAt: servingAt ?? this.servingAt,
      completedAt: completedAt ?? this.completedAt,
      createdAt: createdAt ?? this.createdAt,
      actualServiceSeconds: actualServiceSeconds ?? this.actualServiceSeconds,
      feedback: feedback ?? this.feedback,
    );
  }
}

File: backend/src/models/Token.js
'use strict';

const mongoose = require('mongoose');

/**
 * TOKEN STATUSES:
 *  WAITING   — In queue, waiting to be called
 *  CALLED    — Counter called this token; customer has noShowTimeoutSeconds to arrive
 *  SERVING   — Customer is currently at the counter
 *  COMPLETED — Service successfully delivered
 *  SKIPPED   — Admin skipped this token
 *  CANCELLED — Customer cancelled the token
 *  EXPIRED   — Customer did not arrive in time after being called (no-show)
 */
const TOKEN_STATUSES = ['WAITING', 'CALLED', 'SERVING', 'COMPLETED', 'SKIPPED', 'CANCELLED', 'EXPIRED'];

const tokenSchema = new mongoose.Schema(
  {
    // Human-readable token code, e.g. A-047
    tokenCode: {
      type: String,
      required: true,
      trim: true,
    },
    // Sequential number within the queue for this service on this day
    tokenNumber: {
      type: Number,
      required: true,
      min: 1,
    },
    userId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: [true, 'User reference is required'],
    },
    centerId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'ServiceCenter',
      required: [true, 'Service center reference is required'],
    },
    serviceId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Service',
      required: [true, 'Service reference is required'],
    },
    // Counter assigned when token is called/serving
    counterId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Counter',
      default: null,
    },
    // Staff/Operator who served, called, or completed this token
    servedBy: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      default: null,
    },
    status: {
      type: String,
      enum: TOKEN_STATUSES,
      default: 'WAITING',
    },
    // Position in queue at the time of joining (not live position)
    initialPosition: {
      type: Number,
      min: 1,
    },
    // Live queue position — updated by queueService as tokens are served
    currentPosition: {
      type: Number,
      min: 0,
      default: null,
    },
    // Estimated wait in minutes at time of joining
    waitEstimateMinutes: {
      type: Number,
      default: null,
    },
    // QR code data (signed compact payload string to be encoded as QR)
    qrData: {
      type: String,
    },
    // Current active QR nonce (jti) — used for one-time replay protection.
    // Stored and checked atomically during verify-qr. Cleared when token becomes terminal.
    qrNonce: {
      type: String,
      default: null,
    },
    // When the current qrNonce was issued (for audit and freshness reference)
    qrIssuedAt: {
      type: Date,
      default: null,
    },
    // Whether the QR has been successfully consumed (one-time check-in enforced)
    qrConsumed: {
      type: Boolean,
      default: false,
    },
    // Notification preferences captured at token creation time
    notifyApp: {
      type: Boolean,
      default: true,
    },
    notifySms: {
      type: Boolean,
      default: false,
    },
    // Intake channel attribution (Phase D)
    channel: {
      type: String,
      enum: ['WEB', 'MOBILE', 'QR', 'WHATSAPP', 'SMS', 'TELEGRAM'],
      default: 'WEB',
    },
    channelMetadata: {
      externalMessageId: { type: String, default: null },
      externalUserId: { type: String, default: null },
    },
    // Timestamps for lifecycle events
    calledAt: { type: Date, default: null },
    servingAt: { type: Date, default: null },
    completedAt: { type: Date, default: null },
    // Actual service duration in seconds (set on completion)
    actualServiceSeconds: {
      type: Number,
      default: null,
    },
    // Feedback
    feedback: {
      rating: { type: Number, min: 1, max: 5, default: null },
      comment: { type: String, maxlength: 500, default: null },
      submittedAt: { type: Date, default: null },
    },
  },
  {
    timestamps: true,
    toJSON: { virtuals: true },
  }
);

// ─── Virtuals ─────────────────────────────────────
tokenSchema.virtual('isActive').get(function () {
  return ['WAITING', 'CALLED', 'SERVING'].includes(this.status);
});

// ─── Indexes ──────────────────────────────────────
// Primary query patterns:
tokenSchema.index({ centerId: 1, serviceId: 1, status: 1 });
tokenSchema.index({ userId: 1, status: 1 });
tokenSchema.index({ userId: 1, createdAt: -1 });
tokenSchema.index({ centerId: 1, createdAt: -1 });
tokenSchema.index({ centerId: 1, status: 1, createdAt: -1 });
tokenSchema.index({ centerId: 1, completedAt: -1 });
tokenSchema.index({ counterId: 1, status: 1 });
// Tier 3 / Feature 1 — Context-Aware EWT: bounded service-time history lookup
// for a single service queue (center + service + completion time). Additive only.
tokenSchema.index({ centerId: 1, serviceId: 1, completedAt: -1 });
// Database-level concurrency guarantee: exactly one active token per user per service
tokenSchema.index(
  { userId: 1, centerId: 1, serviceId: 1 },
  {
    unique: true,
    partialFilterExpression: {
      status: { $in: ['WAITING', 'CALLED', 'SERVING'] },
    },
    name: 'unique_active_user_token_per_service',
  }
);

const Token = mongoose.model('Token', tokenSchema);

module.exports = { Token, TOKEN_STATUSES };
